[Aptos Framework] [Staking] Switch from remaining lockup to recurring lockup model

movekevin opened the pull request:
### Description

In the previous model, rewards is based on remaining lockup and linearly decrease over time as remaining lockup goes to 0 (assuming validator doesn't explicitly renew). This creates more overhead for validators than originally expected, especially for staking services: Validators have to constantly decide between extending their lockup for max rewards and letting their lockup expire without extending so they can withdraw part of their funds. 

This PR introduces a new lockup model where the validators are on a recurring "subscription". While they're still in the active validator set, their stake will continuously get automatically renewed for a fixed lockup duration (e.g. one month). They can request to unlock funds at any time, even when this lockup has not expired, but their stake will not be withdrawable until the end of the current lockup period.

An example validator’s flow looks like below:
1. Alice wants to become a validator. She adds 100 coins for the stake.
2. Alice then requests to join the validator set when her node is ready. Once she’s added to the active her validator set, a one-month lockup is automatically enforced on her total stake.
3. Every epoch, Alice receives rewards based on her validator node’s proposal performance.
4. Alice’s stake lockup is renewed every month starting from her first added stake. This is similar to having your T-mobile subscription automatically renewed when the current billing cycle ends.
5. Alice wants to take 5 coins out for living expenses. She sends a transaction to request withdrawal of 5 coins. At the end of the current 1 month cycle, 5 coins are unlocked and Alice can withdraw those 5 coins. The remaining 95 coins are automatically locked up for the next month (automatic renewal).
6. 25 days into the current one month cycle, a governance proposal is submitted with a voting duration of 7 days. Alice only has 5 days left in her current lockup. This is not enough to guarantee her stake will stay locked for the 7-day voting period. Alice wants to participate in voting on this proposal so she resets her lockup, which locks her stake for 30 days. She can now vote on the governance proposal.

### Test Plan
Unit tests + e2e tests (WIP). We'll also heavily test this in internal testnets, devnet, and AIT3.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/2417)
<!-- Reviewable:end -->

movekevin commented on aptos-move/framework/aptos-framework/sources/timestamp.move:
```diff
@@ -164,11 +164,16 @@ module aptos_framework::timestamp {
 
     #[test_only]
-    public fun update_global_time_for_test(timestamp: u64) acquires CurrentTimeMicroseconds {
+    public fun update_global_time_for_test(timestamp_microsecs: u64) acquires CurrentTimeMicroseconds {
```
This was very confusing before as most modules usually obtain timestamps in seconds, while this takes timestamp in microseconds.

movekevin commented on aptos-move/framework/aptos-framework/sources/timestamp.move:
```diff
@@ -164,11 +164,16 @@ module aptos_framework::timestamp {
+    }
+
+    #[test_only]
+    public fun fast_forward_seconds(timestamp_seconds: u64) acquires CurrentTimeMicroseconds {
```
A lot more direct function to move time forward, which most tests do. Also goes by secs instead of the confusing microseconds.

ikabiljo replied:
timestamp_seconds => duration_seconds?

movekevin commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -752,31 +738,19 @@ module aptos_framework::stake {
         let stake_pool = borrow_global_mut<StakePool>(pool_address);
         let unlocked_stake = coin::extract<AptosCoin>(&mut stake_pool.active, amount);
-
-        // Withdrawal is not allowed if lockup has not expired.
```
This is no longer needed as stake is moved from active to pending_inactive and not inactive so it's not withdrawable. Pending_inactive funds are only moved to inactive in the first epoch after the current lockup expires.

movekevin commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -1013,34 +1009,37 @@ module aptos_framework::stake {
         );
         let rewards_amount = coin::value<AptosCoin>(&rewards);
-        coin::merge(&mut stake_pool.active, rewards);
+        // Merge rewards into active unless the validator is unlocking all of their stake.
```
Minting to both active and pending_inactive is also an option but it can add complexity, especially if the pending_inactive amount is very small (which can round rewards for it down to 0)

movekevin commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -1013,34 +1009,37 @@ module aptos_framework::stake {
+
+        // Pending inactive stake is only fully unlocked and moved into inactive if the current lockup cycle has expired
+        let current_lockup_expiration = stake_pool.locked_until_secs;
+        if (timestamp::now_seconds() >= current_lockup_expiration) {
```
This is a key part of the new flow - stakes can be unlocked anytime, but is still subject to the lockup and doesn't become withdrawable until the lockup has expired.

zekun000 replied:
I'll need to think about this more, it seems not very clean to mixed up on lockup time and epochs

movekevin replied:
This does add some risk to the critical path (epoch transition) but does make the UX a lot better - delegations can easily be built on this for example, as they can request partial unlock and settle via a second call to withdraw when it's ready (remaining lockup time is also very easily obtainable via get_remaining_lockup_secs).

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -650,37 +641,33 @@ module aptos_framework::stake {
-        account: &signer,
-        new_locked_until_secs: u64,
-    ) acquires OwnerCapability, StakePool, StakePoolEvents, ValidatorSetConfiguration {
+    public entry fun increase_lockup(account: &signer)
```
do we still need an entry function for it? I thought it's going to be recurring?

movekevin replied:
This is for "resetting" the lockup that's needed for governance. It's the last point in the flow mentioned in the description:

6. 25 days into the current one month cycle, a governance proposal is submitted with a voting duration of 7 days. Alice only has 5 days left in her current lockup. This is not enough to guarantee her stake will stay locked for the 7-day voting period. Alice wants to participate in voting on this proposal so she resets her lockup, which locks her stake for 30 days. She can now vote on the governance proposal.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -752,31 +738,19 @@ module aptos_framework::stake {
-        } else {
-            coin::merge<AptosCoin>(&mut stake_pool.inactive, unlocked_stake);
-        };
+        coin::merge<AptosCoin>(&mut stake_pool.pending_inactive, unlocked_stake);
```
hmm, so pending_inactive only changes after lockup expires instead of at epoch boundary?

movekevin replied:
Yes. This is the best option for a recurring lockup I think. Otherwise, users would need to time their unlock/withdraw to coincide with the lockup expiration.

davidiw commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -1,20 +1,21 @@
+ * automatically renewed at expiration.
+ * 5. At any point, if the validator operator wants to update consensus key or network/fullnode addresses, they can call
+ * stake::rotate_consensus_key and stake::update_network_and_fullnode_addresses.
+ * 6. Validator can request to unlock their stake at any time. However, their stake will only become withdrawable when
```
A validator

they can withdraw their stake when

davidiw commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -1,20 +1,21 @@
  * or if their stake drops below the min required, they would get removed at the end of the epoch.
  * 8. Validator can always rejoin the validator set by going through steps 2-3 again.
- * 9. Owner can always switch operators by calling Stake::set_operator.
+ * 9. Owner can always switch operators by calling stake::set_operator.
```
An owner

davidiw commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -650,37 +641,33 @@ module aptos_framework::stake {
-        account: &signer,
-        new_locked_until_secs: u64,
-    ) acquires OwnerCapability, StakePool, StakePoolEvents, ValidatorSetConfiguration {
+    public entry fun increase_lockup(account: &signer)
```
This is awkward... maybe we allow the voter to increase this as well?

davidiw commented:
Some other comments from the entirety of the code ... just for posterity sake:

* Can we move configuration to another file or at least separate the file into distinct boundary points of structs and functions?
* Should we move delegated_voter and operator_address into OwnerCapability
* Why is the OwnerCapability both store and key?
* We could probably separate out OwnerCapability from this file if we used the Capability where the store is different than the Capability itself
* Has @zekun000  and you confirmed that the entries for a Validator set and their configuration are defined only at epoch boundaries? e.g., a network address rotation or consensus key rotation in the middle will not affect the active validator set?
* FWIW, a fullnode address rotation can take place immediately, would recommend we just split these into separate functions
* I think you might have identified a new paradigm here... allow only entry points to take signers and validate that you should acquire something that doesn't require signer thereon -- this concept is still not quite clear with _with_cap but it is an interesting step forward
* a little dangerous to not make withdrawal amount abort if too high... any case where we would just want to withdraw a portion? It seems more sane to just say all of it? they can add it back if they really want to...
* is_inactive_validator is undefined from comments but interesting that pending_active makes it ... not inactive... maybe make it offline_validator?
* your loops for pending_inactive and active_validators on update_stake_pool could be further removed to a function that just takes in a vec<Validator...>.
* when creating a new validator set rename active_validators -> next_epoch_validators
* bubble sorts are the best
* you've reimplemented vector::append: https://github.com/aptos-labs/aptos-core/blob/8464a9d61b18b791d20d85aac10d78383246d569/aptos-move/framework/move-stdlib/sources/vector.move#L85
* Might use a continue in the if conditional if the minimum stake is not met, that way you don't have to read the rest of the code

movekevin commented:
> Some other comments from the entirety of the code ... just for posterity sake:
> 
> * Can we move configuration to another file or at least separate the file into distinct boundary points of structs and functions?

We can. ValidatorSetConfigurations and ValidatorSet may be can be refactored into a separate validator_set module for example. It's not the cleanest as validator_set would need to expose friend-only functions that only stake can call, but it does make the code shorter. Can be done in a separate PR.

> * Should we move delegated_voter and operator_address into OwnerCapability
But then how would a StakePool have different accounts holding these 3 roles? Would it need to issue different OwnerCapability structs?

> * Why is the OwnerCapability both store and key?
It's key so that an account can have it directly. It's store so that modules (specifically staking services) can store it and control StakePool.

> * We could probably separate out OwnerCapability from this file if we used the Capability where the store is different than the Capability itself
Perhaps but at the cost of more boilerplate code (getters).

> * Has @zekun000  and you confirmed that the entries for a Validator set and their configuration are defined only at epoch boundaries? e.g., a network address rotation or consensus key rotation in the middle will not affect the active validator set?
I think so but @zekun000 to confirm.

> * FWIW, a fullnode address rotation can take place immediately, would recommend we just split these into separate functions
Makes sense. Can be done in a separate PR.

> * I think you might have identified a new paradigm here... allow only entry points to take signers and validate that you should acquire something that doesn't require signer thereon -- this concept is still not quite clear with _with_cap but it is an interesting step forward
Yeah. An alternative paradigm is to only use signers. Autonomous accounts can store their own SignerCapability so if most modules are deployed to their own accounts, this can work as well.

> * a little dangerous to not make withdrawal amount abort if too high... any case where we would just want to withdraw a portion? It seems more sane to just say all of it? they can add it back if they really want to...
Yes partial withdrawal can make the UX for staking services a lot easier, as they don't need to pull all the funds at once but only do so partially as their users withdraw.

> * is_inactive_validator is undefined from comments but interesting that pending_active makes it ... not inactive... maybe make it offline_validator?
I removed is_inactive_validator. get_validator_state returns all different states and can be used to be more exact.

> * your loops for pending_inactive and active_validators on update_stake_pool could be further removed to a function that just takes in a vec<Validator...>.
Yes. Can be done in a separate PR.

> * when creating a new validator set rename active_validators -> next_epoch_validators
Updated.

> * bubble sorts are the best
Lol. I'll let @ikabiljo decide later if that can be killed. Failed proposer indices might still depend on it.

> * you've reimplemented vector::append: https://github.com/aptos-labs/aptos-core/blob/8464a9d61b18b791d20d85aac10d78383246d569/aptos-move/framework/move-stdlib/sources/vector.move#L85
Why does vector::append also reverse the order of the second vector??? It also automatically destroys the second vector. Seems like a potential waste of gas. We can also remove stake::append in a separate PR (and decide if the reverse call in vector::append makes any sense at all).

> * Might use a continue in the if conditional if the minimum stake is not met, that way you don't have to read the rest of the code
Updated.

ikabiljo commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -1,20 +1,21 @@
- * 7. After exiting, the validator can either explicitly leave the validator set by calling Stake::leave_validator_set
+ * 4. Validate and gain rewards. The stake will automatically be locked up for a fixed duration (set by governance) and
+ * automatically renewed at expiration.
+ * 5. At any point, if the validator operator wants to update consensus key or network/fullnode addresses, they can call
```
clarify whether such update becomes effective immediately, or on epoch change.

ikabiljo commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -328,9 +324,9 @@ module aptos_framework::stake {
     }
 
     /// Return the required min/max lockup durations.
```
update comment

movekevin replied:
Updated

ikabiljo commented on aptos-move/framework/aptos-framework/sources/configs/stake.move:
```diff
@@ -356,17 +352,15 @@ module aptos_framework::stake {
 
     /// Update the min and max lockup duration.
```
comment

movekevin replied:
Updated

ikabiljo reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2417)* status: 0 of 19 files reviewed, 11 unresolved discussions (waiting on @aching, @banool, @gregnazario, @junkil-park, @movekevin, @sitalkedia, and @zekun000)

___
*[`aptos-move/framework/aptos-framework/sources/configs/stake.move` line 947 at r2](https://reviewable.io/reviews/aptos-labs/aptos-core/2417#-N8Ua3_aD0UdGNzC-ZCy:-N8Ua3_aD0UdGNzC-ZCz:b-11n70u) ([raw file](https://github.com/aptos-labs/aptos-core/blob/d5833c2c64619619a68a7aa77bcef1d436bcba37/aptos-move/framework/aptos-framework/sources/configs/stake.move#L947)):*
> ```Move
>             vector::push_back(&mut validator_perf.missed_votes, 0);
> 
>             // Automatically renew a validator's lockup for validators that will still in the validator set in the next
> ```

*... still be in the ...
___
*[`aptos-move/framework/aptos-framework/sources/configs/stake.move` line 952 at r2](https://reviewable.io/reviews/aptos-labs/aptos-core/2417#-N8UaMZn6RjiM9WPU1hp:-N8UaMZn6RjiM9WPU1hq:bsma244) ([raw file](https://github.com/aptos-labs/aptos-core/blob/d5833c2c64619619a68a7aa77bcef1d436bcba37/aptos-move/framework/aptos-framework/sources/configs/stake.move#L952)):*
> ```Move
>             if (stake_pool.locked_until_secs <= timestamp::now_seconds()) {
>                 stake_pool.locked_until_secs =
>                     timestamp::now_seconds() + validator_set_config.recurring_lockup_duration_secs;
> ```

Is there a potential concern with this drifting a bit? i.e. if epoch is 1 day, after a few months lockup can drift off for a few days from what user might expect?


vs using previous stake_pool.locked_until_secs  instead of timestamp::now_seconds() as a base to count lockup from?


<!-- Sent from Reviewable.io -->

ikabiljo commented:
(sorry for the incremental commenting. your change didn't want to load properly on my end, and github UI was breaking, so switched to reviewable for the last comments ... )

movekevin commented:
> > _[Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2417)_ status: 0 of 19 files reviewed, 11 unresolved discussions (waiting on @aching, @banool, @gregnazario, @junkil-park, @movekevin, @sitalkedia, and @zekun000)
> 
> _[`aptos-move/framework/aptos-framework/sources/configs/stake.move` line 947 at r2](https://reviewable.io/reviews/aptos-labs/aptos-core/2417#-N8Ua3_aD0UdGNzC-ZCy:-N8Ua3_aD0UdGNzC-ZCz:b-11n70u) ([raw file](https://github.com/aptos-labs/aptos-core/blob/d5833c2c64619619a68a7aa77bcef1d436bcba37/aptos-move/framework/aptos-framework/sources/configs/stake.move#L947)):_
> 
> > ```move
> >             vector::push_back(&mut validator_perf.missed_votes, 0);
> > 
> >             // Automatically renew a validator's lockup for validators that will still in the validator set in the next
> > ```
> 
> *... still be in the ...

Updated

> _[`aptos-move/framework/aptos-framework/sources/configs/stake.move` line 952 at r2](https://reviewable.io/reviews/aptos-labs/aptos-core/2417#-N8UaMZn6RjiM9WPU1hp:-N8UaMZn6RjiM9WPU1hq:bsma244) ([raw file](https://github.com/aptos-labs/aptos-core/blob/d5833c2c64619619a68a7aa77bcef1d436bcba37/aptos-move/framework/aptos-framework/sources/configs/stake.move#L952)):_
> 
> > ```move
> >             if (stake_pool.locked_until_secs <= timestamp::now_seconds()) {
> >                 stake_pool.locked_until_secs =
> >                     timestamp::now_seconds() + validator_set_config.recurring_lockup_duration_secs;
> > ```
> 
> Is there a potential concern with this drifting a bit? i.e. if epoch is 1 day, after a few months lockup can drift off for a few days from what user might expect?
> 
> vs using previous stake_pool.locked_until_secs instead of timestamp::now_seconds() as a base to count lockup from?

Good point. Using stake_pool.locked_until_secs though has the opposite problem - the new lockup might not be exactly 30 days. It'd become 30 days - 1 epoch, which is also equally strange. This might not be too much of a problem though if our epochs are short (e.g. 1 hour).

zekun000 reviewed (commented):
overall looks good to me, I think repurposing the pending_inactive is probably the best solution.

one thing that feels a little weird is that a validator with pending_inactive can leave and re-join (with pending_inactive potentially expired). if it doesn't have minimum stake in active_stake, it'll be kicked out in next epoch due to the unlock. even it's not a big problem, I still hope we can maintain some invariant that new validators only carry active stake
